unpack_unix_test.go

```diff
@@ -0,0 +1,256 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+//go:build unix
+
+package extract_test
+
+import (
+	"context"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/go-extract"
+)
+
+func TestUnpackWithPreserveFileAttributes(t *testing.T) {
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var (
+				ctx = context.Background()
+				dst = t.TempDir()
+				src = asIoReader(t, tc.packer(t, tc.contents))
+				cfg = extract.NewConfig()
+			)
+			if err := extract.Unpack(ctx, dst, src, cfg); err != nil {
+				t.Fatalf("error unpacking archive: %v", err)
+			}
+			for _, c := range tc.contents {
+				path := filepath.Join(dst, c.Name)
+				stat, err := os.Lstat(path)
+				if err != nil {
+					t.Fatalf("error getting file stats: %v", err)
+				}
+				if !(c.Mode&fs.ModeSymlink != 0) { // skip symlink checks
+					if stat.Mode().Perm() != c.Mode.Perm() {
+						t.Fatalf("expected file mode %v, got %v, file %s", c.Mode.Perm(), stat.Mode().Perm(), c.Name)
+					}
+				}
+				if tc.doesNotSupportModTime {
+					continue
+				}
+				modTimeDiff := abs(stat.ModTime().UnixNano() - c.ModTime.UnixNano())
+				if modTimeDiff >= int64(time.Microsecond) {
+					t.Fatalf("expected mod time %v, got %v, file %s, diff %v", c.ModTime, stat.ModTime(), c.Name, modTimeDiff)
+				}
+			}
+		})
+	}
+}
+
+func TestUnpackWithPreserveOwnershipAsNonRoot(t *testing.T) {
+
+	if os.Getuid() == 0 {
+		t.Skip("test requires non-root privileges")
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var (
+				ctx = context.Background()
+				dst = t.TempDir()
+				src = asIoReader(t, tc.packer(t, tc.contents))
+				cfg = extract.NewConfig(extract.WithPreserveOwner(true))
+			)
+			// compare archive uid/gid with current user
+			archiveEntriesEqualCurrentOwner := tc.doesNotSupportOwner
+
+			// If the archive supports owner information, check if the current user is the owner of the test data
+			if !tc.doesNotSupportOwner && (os.Getuid() != tc.uid || os.Getgid() != tc.gid) {
+				archiveEntriesEqualCurrentOwner = false
+			}
+
+			// Unpack should fail if the user is not root and the uid/gid
+			// in the archive is different from the current user (only
+			// if the archive supports owner information)
+			err := extract.Unpack(ctx, dst, src, cfg)
+
+			// chown will only fail if the user is not root and the uid/gid is different
+			if archiveEntriesEqualCurrentOwner && err != nil {
+				t.Fatalf("error unpacking archive: %v", err)
+			}
+		})
+	}
+}
+func TestUnpackWithPreserveOwnershipAsRoot(t *testing.T) {
```

Comment:
Related to the comment above, we probably want to make sure that the uid/gid pair of at least one arhive entry in this test case does not match the current process, or it would pass trivially in that case. You can probably add a single new test case to the slice of cases with some "unusual" upd/gid pair...

Reply:
I adressed this comment above in a certain way. `uid`/`gid` can be 32bit (https://github.com/hashicorp/go-extract/pull/106#discussion_r1907290233) and so `-1` is unexpected, but still a valid 32bit integer. Not sure how we can create a more "unusual" ownership. 

However, I think I addressed this request in https://github.com/hashicorp/go-extract/commit/064184a0fdfff8a0f2efbce8d3378092fc0df042 and https://github.com/hashicorp/go-extract/commit/6c8d47a3b219495b7b8e137027246829e677d335

Reply:
Ha, rolling back on ^ `uid=-1` is not applied by `chown` and fails. However, figured out in https://github.com/hashicorp/go-extract/pull/106#discussion_r1907299361 a way to test invalid uid/gid.